edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentProtocolHead.cs

```diff
@@ -121,6 +132,33 @@ public void Dispose()
                           .Build();
         }
 
+        public override Task StoreNotificationAsync(bool notification)
+        {
+            _started = true;
+            return Task.FromResult(true);
+        }
+
+        public override Task<IEnumerable<Message>> ConvertStoredNotificationsToMessagesAsync()
+        {
+            IEnumerable<Message> messages;
+            if (_started)
+            {
+                messages = new[] { new Message(Topic, Payload) };
+            }
+            else
+            {
+                messages = new Message[0];
+            }
+
+            return Task.FromResult(messages);
+        }
+
+        public override Task<IEnumerable<Message>> ConvertNotificationToMessagesAsync(bool notification)
+        {
+            IEnumerable<Message> messages = new[] { new Message(Topic, Payload) };
+            return Task.FromResult(messages);
+        }
+
```

Comment:
I think it is too much logic here in the protocol head class just to be able to handle both situations when the connector is connected or not connected. I feel that this should not be the concern of this class. Can it be implemented the way that this (and the ScopeIdentities) would not care about the connected state - they just call a 'SendMessageWheneverYouCanAsync()' method with the baked message (topic and binary content, just like connector.SendAsync(), and then the Connector would contain the logic that once it is connected, it sends the messages out? In that case we don't need these three methods here and in the ScopeIdentities, and both AuthAgentProtocolHead and ScopeIdentities would have a very narrow code to implement the notification. Also, we could get rid of the base class. 
The connector class could store these messages in a MqttPublishInfo it already uses (so we could get rid of the Message class introduced here), and once it gets connected, it fires out the collected messages.

Reply:
The 3 API is very clear: prepare what message need to be send. 

- `ConvertStoredNotificationsToMessagesAsync` is based on if `sendNotificationWhenConnected`, return the notification message.
- `ConvertNotificationToMessagesAsync`is return the notification message.
- `StoreNotificationAsync` is ask to store state while notification message can't be sent(connection to broker is not established).